cli/bin/semgrep

```diff
@@ -68,25 +68,36 @@ def find_semgrep_core_path():
-    # message showing all the subcommands. So simpler to just remove --experimental.
-    sys.argv.remove("--experimental")
-    # We could have moved the code below in a separate 'osemgrep', like for 'pysemgrep',
-    # but we don't want users to be exposed to another command, so better to hide it.
-    # We expose 'pysemgrep' because osemgrep itself might need to fallback to pysemgrep
-    # and it's better to avoid the possibility of an infinite loop by simply using
-    # a different program name. Morever, in case of big problems, we can always
-    # tell users to run pysemgrep instead of semgrep and be sure they'll get the
-    # old behavior.
-    path = find_semgrep_core_path()
-    # If you call semgrep-core as osemgrep, then we get osemgrep behavior,
-    # see src/main/Main.ml
-    sys.argv[0] = "osemgrep"
-    os.execvp(str(path), sys.argv)
-else:
-    os.execvp("pysemgrep", sys.argv)
+# when python imports a module, i.e. `import semgrep.FOO`, it runs the module's
+# top-level code, which is the code that is not inside any function or class.
+#
+# This means that if we try and run a function from the semgrep package in a subprocess,
+# we would end up calling os.execvp or main(), which causes a bunch of errors. So
+# we need to make sure that we only run the below code when this file is run as an executable.
+#
+# The above explanation is my understanding, but tbh I think python is weirder under the hood
+# with this stuff.
+if __name__ == "__main__":
+    if "--experimental" in sys.argv:
+        #alt: we could  extend osemgrep (and pysemgrep) to accept a new --experimental
+        # flag instead of removing the flag below. However, in osemgrep we do a few
+        # pattern matching of Sys.argv and if we have 'osemgrep --experimental --help'
+        # this would be interpreted as 'osemgrep scan --experimental --help' (because
+        # scan is the default command), and we would not get the simple semgrep --help
+        # message showing all the subcommands. So simpler to just remove --experimental.
+        sys.argv.remove("--experimental")
+        # We could have moved the code below in a separate 'osemgrep', like for 'pysemgrep',
+        # but we don't want users to be exposed to another command, so better to hide it.
+        # We expose 'pysemgrep' because osemgrep itself might need to fallback to pysemgrep
+        # and it's better to avoid the possibility of an infinite loop by simply using
+        # a different program name. Morever, in case of big problems, we can always
+        # tell users to run pysemgrep instead of semgrep and be sure they'll get the
+        # old behavior.
+        path = find_semgrep_core_path()
+        # If you call semgrep-core as osemgrep, then we get osemgrep behavior,
+        # see src/main/Main.ml
+        sys.argv[0] = "osemgrep"
+        os.execvp(str(path), sys.argv)
+    else:
+        import semgrep.__main__
```

Comment:
Oh I didn't see this change in the original PR ... could we restore the execvp("pysemgrep", ...)?
Why does semgrep --test execute the code in this file anyway?

Reply:
And why this was causing an error (I still don't understand the scenario in which we crash and the different code paths).

Reply:
No we can't remove thjs change, it's part of the fix for this issue, and for fixing autocomplete. There's some explanation for why it fixes autocomplete in the closed PR linked above. I'm also somewhat confused, but I think the best explanation I can give is that for testing, we were calling the invoke semgrep function from multiprocess. So Semgrep is called -> enters into the above -> forks and runs semgrep cli -> test command called -> we run `invoke_semgrep` a few times in parallel -> Since python has the GIL, it can't run a function in a separate thread, it has to start a whole new interpreter (or something, I'm not 100% sure) -> new process means we run all top level code -> since semgrep is imported, this means the above semgrep top level code is called (this phase is called bootstrapping) -> os.exec is called during bootstrapping, before the function that is this process's main is run, and this replaces the mid bootstrapped process before it's had a chance to run -> python does not like this.

The __name__ check stops this from happening, but I'm not sure why the import semgrep.__main__ is needed too. Just that import part fixes autocomplete too (I believe it's something to do with passing environment variables). 

The above is really just my best guess, and I think it's mostly right, but this bug was weird and fix was weird. I think the main reason is that python just really isn't good at multiprocessing, and there's a lot of foot guns

Reply:
Then we need to add your big comment in the script.
Also what is autocomplete? I don't think it's supported by osemgrep anyway, so we could get rid of it.

Reply:
I'll add some of your comment in my PR